File: rust/crates/mapper/tests/normalize.rs
//! Port of `providers/normalize.test.ts` + wire-conversion golden tests.

use tt_mapper::{
    assemble_mapped_status, compute_row_current_serial, to_mapped_station, to_wire_status,
    AssembleOptions, KnownTrain, ProviderStationRow,
};

fn row(overrides: Option<&dyn Fn(&mut ProviderStationRow)>) -> ProviderStationRow {
    let mut row = ProviderStationRow {
        station_code: "NDLS".to_string(),
        station_name: "New Delhi".to_string(),
        scheduled_arrival: Some("08:00".to_string()),
        actual_arrival: Some("08:20".to_string()),
        scheduled_departure: Some("08:05".to_string()),
        actual_departure: None,
        has_departed: None,
        delay_minutes: None,
        distance: Some(391),
        platform: Some("5".to_string()),
        halt_minutes: Some(5),
        day: Some(1),
    };
    if let Some(f) = overrides {
        f(&mut row);
    }
    row
}

fn station_row(code: &str) -> ProviderStationRow {
    let mut r = row(None);
    r.station_code = code.to_string();
    r
}

#[test]
fn compute_row_current_serial_returns_1_based_position() {
    let rows = [station_row("ADI"), station_row("NDLS")];
    assert_eq!(compute_row_current_serial(&rows, Some("NDLS")), 2);
}

#[test]
fn compute_row_current_serial_returns_0_when_no_match() {
    assert_eq!(compute_row_current_serial(&[row(None)], Some("BPL")), 0);
}

#[test]
fn compute_row_current_serial_returns_0_for_null_code() {
    assert_eq!(compute_row_current_serial(&[row(None)], None), 0);
}

#[test]
fn to_mapped_station_maps_a_fully_populated_row() {
    let mapped = to_mapped_station(&row(None), 0, Some("NDLS"), 1);
    assert_eq!(mapped.station_code, "NDLS");
    assert_eq!(mapped.station_name, "New Delhi");
    assert_eq!(mapped.scheduled_arrival.as_deref(), Some("08:00"));
    assert_eq!(mapped.actual_arrival.as_deref(), Some("08:20"));
    assert_eq!(mapped.scheduled_departure.as_deref(), Some("08:05"));
    assert_eq!(mapped.actual_departure, None);
    assert_eq!(mapped.delay_minutes, Some(20));
    assert_eq!(mapped.distance_from_source, Some(391));
    assert_eq!(mapped.platform.as_deref(), Some("5"));
    assert_eq!(mapped.halt_minutes, Some(5));
    assert!(!mapped.has_departed);
    assert!(mapped.is_current);
    assert_eq!(mapped.day, 1);
}

#[test]
fn to_mapped_station_applies_defaults_for_missing_fields() {
    let mut r = row(None);
    r.distance = None;
    let mapped = to_mapped_station(&r, 5, None, 0);
    assert_eq!(mapped.station_code, "NDLS");
    assert_eq!(mapped.delay_minutes, Some(20));
    assert_eq!(mapped.distance_from_source, None);
    assert!(!mapped.has_departed);
    assert!(!mapped.is_current);
    assert_eq!(mapped.day, 1);
}

#[test]
fn marks_rows_before_the_current_serial_as_departed() {
    let mut r = row(None);
    r.actual_departure = None;
    let mapped = to_mapped_station(&r, 0, Some("NDLS"), 4);
    assert!(mapped.has_departed);
}

#[test]
fn keeps_future_rows_undeparted() {
    let mut r = row(None);
    r.actual_departure = None;
    let mapped = to_mapped_station(&r, 5, Some("NDLS"), 4);
    assert!(!mapped.has_departed);
}

#[test]
fn marks_current_station_departed_only_once_it_has_left() {
    let mut at_station = row(None);
    at_station.actual_departure = None;
    let at_station = to_mapped_station(&at_station, 3, Some("NDLS"), 4);
    assert!(!at_station.has_departed);

    let mut departed = row(None);
    departed.actual_departure = Some("08:05".to_string());
    let departed = to_mapped_station(&departed, 3, Some("NDLS"), 4);
    assert!(departed.has_departed);
}

#[test]
fn honors_explicit_has_departed_flag() {
    let mut r = row(None);
    r.has_departed = Some(true);
    let mapped = to_mapped_station(&r, 5, Some("NDLS"), 4);
    assert!(mapped.has_departed);
}

#[test]
fn uses_explicit_delay_minutes_over_computed_delay() {
    let mut r = row(None);
    r.delay_minutes = Some(99);
    let mapped = to_mapped_station(&r, 0, None, 0);
    assert_eq!(mapped.delay_minutes, Some(99));
}

#[test]
fn falls_back_to_null_for_a_missing_distance() {
    let mut r = row(None);
    r.distance = None;
    let mapped = to_mapped_station(&r, 0, None, 0);
    assert_eq!(mapped.distance_from_source, None);
}

#[test]
fn keeps_a_zero_distance() {
    let mut r = row(None);
    r.distance = Some(0);
    let mapped = to_mapped_station(&r, 0, None, 0);
    assert_eq!(mapped.platform.as_deref(), Some("5"));
    assert_eq!(mapped.distance_from_source, Some(0));
}

fn assemble_rows() -> Vec<ProviderStationRow> {
    let mut source = station_row("ADI");
    source.station_name = "Ahmedabad Jn".to_string();
    source.distance = Some(0);
    let mut current = station_row("NDLS");
    current.station_name = "New Delhi".to_string();
    current.distance = Some(938);
    current.day = Some(2);
    vec![source, current]
}

#[test]
fn assemble_maps_full_status() {
    let status = assemble_mapped_status(
        &assemble_rows(),
        &AssembleOptions {
            train_number: "12301".to_string(),
            departure_date: "20260805".to_string(),
            known_train: Some(KnownTrain {
                number: "12301".to_string(),
                name: "Howrah Rajdhani Express".to_string(),
            }),
            current_station_code: Some("NDLS".to_string()),
            status_message: Some("Running <b>on time</b>".to_string()),
            last_updated: Some("2026-08-05T12:00:00Z".to_string()),
        },
    );
    assert_eq!(status.train_number, "12301");
    assert_eq!(status.train_name, "Howrah Rajdhani Express");
    assert_eq!(status.departure_date, "20260805");
    assert_eq!(status.source_station_code, "ADI");
    assert_eq!(status.source_station_name, "Ahmedabad Jn");
    assert_eq!(status.destination_station_code, "NDLS");
    assert_eq!(status.destination_station_name, "New Delhi");
    assert_eq!(status.current_station_code.as_deref(), Some("NDLS"));
    assert_eq!(status.current_station_name.as_deref(), Some("New Delhi"));
    assert_eq!(status.current_delay_minutes, Some(20));
    assert_eq!(status.status_message.as_deref(), Some("Running on time"));
    assert_eq!(status.last_updated.as_deref(), Some("2026-08-05T12:00:00Z"));
    assert_eq!(status.stations.len(), 2);
    assert!(status.stations[1].is_current);
    assert!(!status.stations[1].has_departed);
    assert_eq!(status.stations[1].day, 2);
    assert!(status.stations[0].has_departed);
}

#[test]
fn assemble_uses_fallback_train_name_when_unknown() {
    let status = assemble_mapped_status(
        &assemble_rows(),
        &AssembleOptions {
            train_number: "12301".to_string(),
            departure_date: "20260805".to_string(),
            ..Default::default()
        },
    );
    assert_eq!(status.train_name, "Train 12301");
}

#[test]
fn assemble_handles_empty_station_list() {
    let status = assemble_mapped_status(
        &[],
        &AssembleOptions {
            train_number: "12301".to_string(),
            departure_date: "20260805".to_string(),
            ..Default::default()
        },
    );
    assert!(status.stations.is_empty());
    assert_eq!(status.source_station_code, "");
    assert_eq!(status.current_station_code, None);
}

#[test]
fn assemble_leaves_current_station_null_when_none_matches() {
    let status = assemble_mapped_status(
        &assemble_rows(),
        &AssembleOptions {
            train_number: "12301".to_string(),
            departure_date: "20260805".to_string(),
            current_station_code: Some("ABC".to_string()),
            ..Default::default()
        },
    );
    assert_eq!(status.current_station_name, None);
    assert_eq!(status.current_delay_minutes, None);
}

#[test]
fn to_wire_status_is_identity_with_explicit_nulls() {
    let status = assemble_mapped_status(
        &assemble_rows(),
        &AssembleOptions {
            train_number: "12301".to_string(),
            departure_date: "20260805".to_string(),
            known_train: Some(KnownTrain {
                number: "12301".to_string(),
                name: "Howrah Rajdhani Express".to_string(),
            }),
            current_station_code: Some("NDLS".to_string()),
            last_updated: Some("2026-08-05T12:00:00Z".to_string()),
            ..Default::default()
        },
    );
    let wire = to_wire_status(&status);
    assert_eq!(wire.train_number, "12301");
    assert_eq!(wire.train_name, "Howrah Rajdhani Express");
    assert_eq!(wire.stations.len(), 2);
    assert_eq!(wire.stations[0].delay_minutes, Some(20));
    assert_eq!(wire.stations[0].actual_departure, None);
    assert!(wire.stations[1].is_current);

    let json = serde_json::to_value(&wire).expect("serialize wire status");
    assert_eq!(
        json.get("status_message"),
        Some(&serde_json::Value::Null),
        "nullish status_message must serialize as explicit null, not be omitted"
    );
    assert_eq!(
        json.pointer("/stations/0/actual_departure"),
        Some(&serde_json::Value::Null),
        "nullish station fields must be explicit null"
    );
}

File: artifacts/api-server/src/lib/providers/normalize.ts
import { calcDelay, stripHtml } from "@workspace/trains-data";
import type { MappedStation, MappedStatus } from "../train-status-mapper";

/**
 * Normalized, provider-agnostic station row. Adapters translate their own
 * payloads into this shape; `toMappedStation`/`assembleMappedStatus` then turn
 * rows into the shared `MappedStatus` contract. This is the DRY seam that
 * keeps every upstream behind a single normalization path.
 */
export interface ProviderStationRow {
  station_code: string;
  station_name: string;
  scheduled_arrival: string | null;
  actual_arrival: string | null;
  scheduled_departure: string | null;
  actual_departure: string | null;
  /**
   * Explicit "already departed" flag, when the upstream tells us directly.
   * When absent it is inferred positionally (rows before the current station
   * are departed; the current station is departed once it has an actual
   * departure time), matching the legacy Paytm mapping.
   */
  has_departed?: boolean;
  delay_minutes?: number | null;
  distance?: number | null;
  platform?: string | null;
  halt_minutes?: number | null;
  day?: number;
}

export interface AssembleOptions {
  trainNumber: string;
  departureDate: string;
  knownTrain: { number: string; name: string } | null;
  /** Code of the station the train is currently at/past, if known. */
  currentStationCode?: string | null;
  statusMessage?: string | null;
  lastUpdated?: string | null;
}

/** 1-based serial of `currentStationCode` within the row list, else 0. */
export function computeRowCurrentSerial(
  rows: ProviderStationRow[],
  currentStationCode: string | null,
): number {
  if (!currentStationCode) return 0;
  const index = rows.findIndex(
    (row) => row.station_code === currentStationCode,
  );
  return index < 0 ? 0 : index + 1;
}

export function toMappedStation(
  row: ProviderStationRow,
  rowIndex: number,
  currentStationCode: string | null,
  currentSerial: number,
): MappedStation {
  const isCurrent = row.station_code === currentStationCode;
  const serial = rowIndex + 1;
  const hasDeparted =
    row.has_departed !== undefined
      ? row.has_departed
      : serial < currentSerial || (isCurrent && !!row.actual_departure);

  const scheduledArrival = row.scheduled_arrival;
  const scheduledDeparture = row.scheduled_departure;
  const distance = row.distance ?? null;

  return {
    station_code: row.station_code,
    station_name: row.station_name,
    scheduled_arrival: scheduledArrival,
    actual_arrival: row.actual_arrival,
    scheduled_departure: scheduledDeparture,
    actual_departure: row.actual_departure,
    delay_minutes:
      row.delay_minutes !== undefined && row.delay_minutes !== null
        ? row.delay_minutes
        : calcDelay(scheduledArrival, row.actual_arrival),
    distance_from_source:
      distance !== null && Number.isFinite(distance) ? distance : null,
    platform: row.platform ?? null,
    halt_minutes: row.halt_minutes ?? null,
    has_departed: hasDeparted,
    is_current: isCurrent,
    day: row.day ?? 1,
  };
}

/** Turn normalized rows into the shared `MappedStatus` contract. */
export function assembleMappedStatus(
  rows: ProviderStationRow[],
  options: AssembleOptions,
): MappedStatus {
  const currentStationCode = options.currentStationCode ?? null;
  const currentSerial = computeRowCurrentSerial(rows, currentStationCode);
  const stations = rows.map((row, index) =>
    toMappedStation(row, index, currentStationCode, currentSerial),
  );

  const currentStation = stations.find((station) => station.is_current);
  const firstStation = stations[0];
  const lastStation = stations[stations.length - 1];

  return {
    train_number: options.trainNumber,
    train_name: options.knownTrain
      ? options.knownTrain.name
      : `Train ${options.trainNumber}`,
    departure_date: options.departureDate,
    source_station_code: firstStation?.station_code ?? "",
    source_station_name: firstStation?.station_name ?? "",
    destination_station_code: lastStation?.station_code ?? "",
    destination_station_name: lastStation?.station_name ?? "",
    current_station_code: currentStationCode,
    current_station_name: currentStation?.station_name ?? null,
    current_delay_minutes: currentStation?.delay_minutes ?? null,
    status_message: stripHtml(options.statusMessage ?? null),
    last_updated: options.lastUpdated ?? null,
    provider: "",
    stations,
  };
}
